### PR TITLE
HttpStatusPush GithubStatusPush and StashStatusPush port to reporter API

### PR DESCRIPTION
**Pull request by tardyp**
stash, github, gitlab, are all very similar. We make HttpStatusPushBase class, and create 3 classes for sending status update via http

TODO:
* [x] Unit tests
* [x] Factorize in a base class -> HttpStatusPush

### CODE REVIEW

File: master/buildbot/reporters/http.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+from buildbot.util import service
+
+from twisted.internet import defer
+from twisted.python import log
+
+# use the 'requests' lib: http://python-requests.org
+try:
+    import txrequests
+    assert txrequests
+    import requests
```
Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: I thought `txrequests` **depends on** `requests`, is it not the case?

File: master/buildbot/reporters/http.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+from buildbot.reporters import utils
+from buildbot.util import service
+
+from twisted.internet import defer
+from twisted.python import log
+
+# use the 'requests' lib: http://python-requests.org
+try:
+    import txrequests
+    assert txrequests
```
Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: What exactly this line is supposed to do?  If the import is successful, then it's gonna be OK, if not, then `ImportError` would be thrown, AFAIU.

File: master/buildbot/reporters/stash.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+from buildbot.reporters import http
+from buildbot.reporters import utils
+
+from buildbot.process.results import SUCCESS
+from twisted.internet import defer
```
Comment (sa2ajj) [inline master/buildbot/reporters/stash.py]: I thoutht we'd put imports from `twisted` before imports from `buildbot`...

Comment (tardyp) [inline master/buildbot/reporters/stash.py]: I can never remember the convention. As long as it is not checked :)

File: master/buildbot/reporters/http.py (hunk 1 above)
Comment (tardyp) [inline master/buildbot/reporters/http.py]: I actually copied the code from steps/http.py without any more thought

Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: There's a room to save space :smile:

File: master/buildbot/reporters/stash.py (hunk 3 above)
Comment (sa2ajj) [inline master/buildbot/reporters/stash.py]: I understand this very well :smile:

Comment (tardyp) [inline master/buildbot/reporters/stash.py]: actually on all my open files, twisted is after buildbot

Comment (sa2ajj) [inline master/buildbot/reporters/stash.py]: This is something I'd like to submit a patch for for a very long time...

File: master/buildbot/reporters/github.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from buildbot.process.results import EXCEPTION
+from buildbot.process.results import FAILURE
+from buildbot.process.results import RETRY
+from buildbot.process.results import SKIPPED
+from buildbot.process.results import SUCCESS
+from buildbot.process.results import WARNINGS
+from twisted.internet import defer
+from twisted.python import log
+
+HOSTED_BASE_URL = 'https://api.github.com'
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: [nit] missing empty line beforehand (pep-8 requires 2 empty lines in this case)

File: master/buildbot/reporters/github.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    name = "GithubStatusPush"
+    neededDetails = dict(wantProperties=True)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, token,
+                        startDescription=None, endDescription=None,
+                        context=None, baseURL=None, verbose=False, **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(self, **kwargs)
+
+        self.context = context or Interpolate('buildbot/%(prop:buildername)s')
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: should the `buildbot` part be hardcoded here?  what about making it a parameter of the service?

Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: right, on the second thought: you just overwrite `context`

File: master/buildbot/reporters/github.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        if description is not None:
+            payload['description'] = description
+
+        if target_url is not None:
+            payload['target_url'] = target_url
+
+        if context is not None:
+            payload['context'] = context
+
+        return self.session.post("/".join(
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: should this also be '/' (as not "/")?

File: master/buildbot/reporters/github.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        else:
+            state = 'pending'
+            description = yield props.render(self.startDescription)
+
+        context = yield props.render(self.context)
+
+        sourcestamps = build['buildset']['sourcestamps']
+        project = sourcestamps[0]['project']
+
+        if not project:
```
Comment (sa2ajj) [inline master/buildbot/reporters/github.py]: i don't know if it makes a lot of sense, but why not change the condition and swap the branches?

File: master/buildbot/reporters/http.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+from future.utils import iteritems
+
```
Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: unnecessary empty line

**Comment by sa2ajj** [comment]
If it's good to go, I think commits should not have `WIP` in their subjects.

File: master/buildbot/reporters/http.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+            self.buildStarted,
+            ('builds', None, 'started'))
+
+    def stopService(self):
+        self._buildCompleteConsumer.stopConsuming()
+        self._buildStartedConsumer.stopConsuming()
+        self.session.close()
+
+    def buildStarted(self, key, build):
+        return self.getMoreInfoAndSend(build)
```
Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: I'd add `getMoreInfoAndSend` here as an abstract method: that'd help to catch problems when config is loaded.

Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: Well, again I should have read a bit further :/

File: master/buildbot/reporters/http.py [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    def filterBuilds(self, build):
+        if self.builders is not None:
+            return build['builder']['name'] in self.builders
+        return True
+
+    @defer.inlineCallbacks
+    def getMoreInfoAndSend(self, build):
+        yield utils.getDetailsForBuild(self.master, build, **self.neededDetails)
+        if self.filterBuilds(build):
+            yield self.send(build)
```
Comment (sa2ajj) [inline master/buildbot/reporters/http.py]: But then my comment applies to `send`

File: master/buildbot/reporters/stash.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+class StashStatusPush(http.HttpStatusPushBase):
+    name = "StashStatusPush"
+
+    @defer.inlineCallbacks
+    def reconfigService(self, base_url, user, password, **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(self, **kwargs)
+        if not base_url.endswith('/'):
+            base_url += '/'
+        self.base_url = '%srest/build-status/1.0/commits/' % (base_url, )
```
Comment (sa2ajj) [inline master/buildbot/reporters/stash.py]: extra space after `,`

File: master/docs/manual/cfg-reporters.rst [hunk 12]
```diff
@@ -431,7 +427,7 @@ List of responsible users
 .. index:: IRC
 
 IRC Bot
-~~~~~~~~~~~~~~~~~~~~~~
+~~~~~~~
 
 
 The :bb:reporter:`IRC` status target creates an IRC bot which will attach to certain channels and be available for status queries.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-reporters.rst]: `status target` --> `reporter` ?

File: master/docs/relnotes/index.rst [hunk 13]
```diff
@@ -36,6 +39,9 @@ Deprecations, Removals, and Non-Compatible Changes
 
 * Deprecated ``workdir`` was removed, ``builddir`` property should be used instead.
 
+* :class:`~buildbot.status.status_push.StatusPush` has been removed from buildbot.
+    please use the much simpler :class:`~buildbot.reporters.http.HttpStatusPush` instead.
```
Comment (sa2ajj) [inline master/docs/relnotes/index.rst]: _P_lease

**Comment by sa2ajj** [comment]
LGTM otherwise

File: master/buildbot/reporters/github.py (hunk 4 above)
Comment (tardyp) [inline master/buildbot/reporters/github.py]: flake8 wouldn't mind, but anyway done

File: master/buildbot/reporters/github.py (hunk 6 above)
Comment (tardyp) [inline master/buildbot/reporters/github.py]: done

**Comment by tardyp** [comment]
I should have answered all comments

**Comment by tardyp** [comment]
rebased. would apreciate a :+1:

**Comment by sa2ajj** [comment]
Any particular reason you want to keep `WIP` in `WIP: stash port to reporter api`?

**Comment by tardyp** [comment]
well, because its a wip commit. I can squash everything

**Pull request merged**